File: app-webui/protocol.go
package appbackend

import (
	"encoding/json"
	"time"

	"github.com/ingot-agent/sdk/interaction"
)

// Attachment is the asset-first browser attachment DTO.
type Attachment struct {
	Kind     string `json:"kind"`
	MIMEType string `json:"mimeType,omitempty"`
	Name     string `json:"name,omitempty"`
	AssetID  string `json:"assetId"`
}

// ErrorResponse is the stable HTTP error envelope.
type ErrorResponse struct {
	Error ErrorDetail `json:"error"`
}

// WorkspaceBrowseEntry is one selectable subdirectory inside a Workspace
// directory picker. Path is the full absolute host path, so the browser never
// has to join OS-specific path separators itself.
type WorkspaceBrowseEntry struct {
	Name string `json:"name"`
	Path string `json:"path"`
}

// WorkspaceBrowse is the response of the host directory picker. Path is the
// absolute directory being listed; Parent is its parent directory (empty at a
// filesystem root); Roots contains other selectable filesystem roots when the
// current directory is a root; Directories holds the selectable subdirectories
// in sorted order.
type WorkspaceBrowse struct {
	Path        string                 `json:"path"`
	Parent      string                 `json:"parent,omitempty"`
	Roots       []WorkspaceBrowseEntry `json:"roots,omitempty"`
	Directories []WorkspaceBrowseEntry `json:"directories"`
}

// ErrorDetail describes one HTTP API error.
type ErrorDetail struct {
	Code    string `json:"code"`
	Message string `json:"message"`
}

// AgentCapabilities reports independently available execution modes.
type AgentCapabilities struct {
	Run    bool `json:"run"`
	Stream bool `json:"stream"`
}

// AgentState is the agent portion of the bootstrap snapshot.
type AgentState struct {
	Capabilities AgentCapabilities `json:"capabilities"`
}

// AssetState describes the optional browser asset transport.
type AssetState struct {
	Available bool  `json:"available"`
	MaxBytes  int64 `json:"maxBytes"`
}

// Session is the Web projection of session metadata combined with its
// assigned Workspace. Workspace is the root of the session-scoped local
// workspace binding and is empty when the session is not (yet) bound. This is
// an Application-level DTO, not part of the SDK session.Metadata contract.
type Session struct {
	ID         string     `json:"id"`
	Title      string     `json:"title"`
	Workspace  string     `json:"workspace,omitempty"`
	CreatedAt  time.Time  `json:"createdAt"`
	UpdatedAt  time.Time  `json:"updatedAt"`
	ArchivedAt *time.Time `json:"archivedAt,omitempty"`
}

// TurnSnapshot is the recoverable projection of a running turn.
type TurnSnapshot struct {
	ID        string `json:"id"`
	SessionID string `json:"sessionId"`
	Revision  uint64 `json:"revision"`
	Reasoning string `json:"reasoning"`
	Output    string `json:"output"`
}

// StateSnapshot bootstraps all process-local Web state. Session history is
// intentionally fetched separately.
type StateSnapshot struct {
	Cursor               uint64                `json:"cursor"`
	Agent                AgentState            `json:"agent"`
	Assets               AssetState            `json:"assets"`
	Sessions             []Session             `json:"sessions"`
	Operations           []OperationDefinition `json:"operations"`
	Turns                []TurnSnapshot        `json:"turns"`
	OperationInvocations []OperationSnapshot   `json:"operationInvocations"`
	Interactions         []PendingInteraction  `json:"interactions"`
	InteractionStates    []InteractionState    `json:"interactionStates"`
}

// InteractionSubmission carries raw browser values for host-side validation.
type InteractionSubmission struct {
	Values map[string]json.RawMessage `json:"values"`
}

// PendingInteraction is a caller-owned pending request snapshot.
type PendingInteraction struct {
	Scope   *Scope             `json:"scope,omitempty"`
	ID      string             `json:"id"`
	Name    string             `json:"name"`
	Level   string             `json:"level,omitempty"`
	Message string             `json:"description,omitempty"`
	Fields  []InteractionField `json:"fields"`
}

// InteractionField is the browser representation of an SDK request field.
// Object fields carry their members in Fields; list fields carry one element
// descriptor in Element. Both nest arbitrarily deep, so a renderer must
// recurse instead of assuming a flat form.
type InteractionField struct {
	Name        string              `json:"name"`
	Label       string              `json:"label,omitempty"`
	Description string              `json:"description,omitempty"`
	Kind        string              `json:"kind"`
	Required    bool                `json:"required"`
	Sensitive   bool                `json:"sensitive"`
	HasDefault  bool                `json:"hasDefault"`
	Default     any                 `json:"default,omitempty"`
	Options     []InteractionOption `json:"options,omitempty"`
	Fields      []InteractionField  `json:"fields,omitempty"`
	Element     *InteractionField   `json:"element,omitempty"`
}

// InteractionOption is one ordered choice.
type InteractionOption struct {
	Value       string `json:"value"`
	Label       string `json:"label,omitempty"`
	Description string `json:"description,omitempty"`
}

// InteractionState is the current replaceable state projection.
type InteractionState struct {
	Scope       *Scope                  `json:"scope,omitempty"`
	ID          string                  `json:"id"`
	Name        string                  `json:"name"`
	Level       string                  `json:"level,omitempty"`
	Description string                  `json:"description,omitempty"`
	Values      []InteractionStateEntry `json:"values"`
}

// InteractionStateEntry is one projected state value.
type InteractionStateEntry struct {
	Name        string `json:"name"`
	Label       string `json:"label,omitempty"`
	Description string `json:"description,omitempty"`
	Value       any    `json:"value"`
}

// InteractionHost implements the SDK channel and exposes Web settlement and
// authoritative snapshots.
type InteractionHost interface {
	interaction.Channel
	Pending() []PendingInteraction
	States() []InteractionState
	Respond(string, InteractionSubmission) error
	Scoped(Scope) interaction.Channel
}

// OperationDefinition is an immutable operation's public JSON contract. ID is
// the host-generated internal identity used for invocation; Name is the
// Plugin-chosen display identity and may be duplicated across Plugins. Group is
// a Plugin-chosen presentation hint that may be empty ("ungrouped").
type OperationDefinition struct {
	ID           string          `json:"id"`
	Name         string          `json:"name"`
	Description  string          `json:"description"`
	Group        string          `json:"group,omitempty"`
	InputSchema  json.RawMessage `json:"inputSchema"`
	OutputSchema json.RawMessage `json:"outputSchema"`
}

// OperationSnapshot preserves running calls and a bounded set of final results.
// OperationID addresses the invoked operation; Name is its display identity.
type OperationSnapshot struct {
	ID          string           `json:"id"`
	OperationID string           `json:"operationId"`
	Name        string           `json:"name"`
	SessionID   string           `json:"sessionId,omitempty"`
	Status      string           `json:"status"`
	Result      *OperationResult `json:"result,omitempty"`
	Error       *ErrorDetail     `json:"error,omitempty"`
}

// OperationResult is the validated machine-readable output of an operation.
type OperationResult struct {
	Output json.RawMessage `json:"output"`
}

// Runtime contains the host-side process state shared with the HTTP component.
type Runtime interface {
	Events() EventHub
	Interactions() InteractionHost
}

File: app-webui/web/src/protocol.ts
export type Status = 'running' | 'succeeded' | 'failed' | 'canceled'
export interface Session {
  id: string
  title: string
  workspace?: string
  createdAt: string
  updatedAt: string
  archivedAt?: string
}
export interface Part {
  kind: string
  text?: string
  mimeType?: string
  name?: string
  source?: { kind: string; data?: string; uri?: string; assetId?: string }
}
export interface ToolCall { id: string; name: string; arguments: unknown }
export interface Message {
  role: string
  content: Part[]
  name?: string
  toolCallId?: string
  toolCalls?: ToolCall[]
}
export interface Attachment { kind: string; mimeType?: string; name?: string; assetId: string }
export interface Scope {
  agent?: { sessionId?: string; turnId?: string; roundIndex?: number; toolCallId?: string }
  operation?: { invocationId: string }
}
export interface ErrorDetail { code: string; message: string }
export interface Outcome {
  status: Status
  durationNs: number
  accounting: {
    rounds: number
    modelInvocations: number
    toolCalls: number
    usage: { inputTokens: number; outputTokens: number; totalTokens: number; coverage: string }
    models?: unknown[]
  }
  failure?: { stage: string; roundIndex?: number; toolCallId?: string }
}
export interface Turn {
  id: string
  sessionId: string
  revision: number
  reasoning: string
  output: string
}
export type TurnBlock =
  | { id: string; kind: 'output' | 'reasoning'; text: string; boundary: number }
  | { id: string; kind: 'tool'; call: ToolCall; status: string; content?: Part[]; error?: string }
  | { id: string; kind: 'interaction'; interactionId: string }
export interface LiveTurn extends Turn {
  status: Status
  blocks?: TurnBlock[]
  streamBoundary?: number
  sdkTurnId?: string
  historyEnd?: number
  stopping?: boolean
  reconciled?: boolean
  outcome?: Outcome
  result?: { output: Part[] }
  error?: ErrorDetail
}
export interface InteractionField {
  name: string
  label?: string
  description?: string
  kind: 'string' | 'integer' | 'number' | 'boolean' | 'choice' | 'multichoice' | 'object' | 'list'
  required: boolean
  sensitive: boolean
  hasDefault: boolean
  default?: unknown
  options?: { value: string; label?: string; description?: string }[]
  // Object fields carry their members; list fields carry one element
  // descriptor. Both nest, so renderers must recurse.
  fields?: InteractionField[]
  element?: InteractionField
}
export interface Interaction {
  id: string
  name: string
  scope?: Scope
  description?: string
  level?: string
  fields: InteractionField[]
}
export interface InteractionState {
  id: string
  name: string
  scope?: Scope
  level?: string
  description?: string
  values: { name: string; label?: string; description?: string; value: unknown }[]
}
export interface Schema {
  type?: string | string[]
  properties?: Record<string, Schema>
  required?: string[]
  enum?: unknown[]
  title?: string
  description?: string
  default?: unknown
  [key: string]: unknown
}
export interface Operation {
  id: string
  name: string
  description: string
  group?: string
  inputSchema: Schema
  outputSchema: Schema
}
export interface OperationInvocation {
  id: string
  operationId: string
  name: string
  sessionId?: string
  status: Status
  result?: { output: unknown }
  error?: ErrorDetail
}
export interface Snapshot {
  cursor: number
  agent: { capabilities: { run: boolean; stream: boolean } }
  assets?: { available: boolean; maxBytes: number }
  sessions: Session[]
  turns: Turn[]
  interactions: Interaction[]
  interactionStates: InteractionState[]
  operations: Operation[]
  operationInvocations: OperationInvocation[]
}
// Observation and Operation payloads deliberately include arbitrary plugin JSON.
export interface WebEvent { type: string; scope?: Scope; data: Record<string, any> }
export interface TraceEvent extends WebEvent { cursor: number }
export interface Notice { id: number; message: string; level: string; scope?: Scope }
